File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
         int n=nums.size();
        vector<int> dp(n+1,INT_MAX);
        dp[0]=INT_MIN;

        for(int i=0;i<n;i++)
        {
            int idx=upper_bound(dp.begin(),dp.end(),nums[i])-dp.begin();
            //Important condition
            if(nums[i]>dp[idx-1] && nums[i]<dp[idx])
            {
                dp[idx]=nums[i];
            }
        }
        for(int i=n;i>=0;i--)
            if(dp[i]!=INT_MAX)
                return i;
        return 1;
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int n = nums.length;
        int dp[][] = new int[n][n+1];
        for(int i=0;i<n;i++)
            Arrays.fill(dp[i],-1);
        return lis(0,-1,nums,dp);
    }

    int lis(int index,int prevIndex , int[] nums,int[][] dp)
    {
        if(index==nums.length)
            return 0;
        if(dp[index][prevIndex+1]!=-1)
            return dp[index][prevIndex+1];
            
        int skip = lis(index+1,prevIndex,nums,dp);
        int choose = -1;

        if(prevIndex == -1 || nums[index] > nums[prevIndex])
            choose = 1+ lis(index+1, index,nums,dp);

        return dp[index][prevIndex+1]=Math.max(skip,choose);
    }
}